### PR TITLE
PERF: fix name resolution performance in large modules

### LINKED ISSUE #6522: Analysis of .rs files is incredibly slow
## Environment

* **IntelliJ Rust plugin version:** 0.3.137.3543-203
* **Rust toolchain version:** 1.47.0 (18bf6b4f0 2020-10-07) x86_64-apple-darwin
* **IDE name and version:** CLion 2020.3 (CL-203.5981.166)
* **Operating system:** macOS 10.15.1
* **Macro expansion engine:** new
* **Name resolution engine:** old

## Problem description

Yesterday (even with the -nightly rust plugin), after working on some major changes to a very large Rust codebase, the analysis engine has become incredibly slow.  It now seems to take >5mins to analyze a file.

My project is quite large with over 250k lines of rust (the majority of this is bindgen-generated code):

```
$ loc   
--------------------------------------------------------------------------------
 Language             Files        Lines        Blank      Comment         Code
--------------------------------------------------------------------------------
 Rust                   136       250199         2633         2702       244864
 Markdown                19          903          253            0          650
 Toml                    20          701           94          120          487
 C/C++ Header             4          321            5            0          316
 YAML                     1          135            2            2          131
 C                        1          150           29            5          116
 Plain Text              13           40            0            0           40
 Bourne Shell             3           41           13            5           23
 Makefile                 1           20            7            2           11
--------------------------------------------------------------------------------
 Total                  198       252510         3036         2836       246638
--------------------------------------------------------------------------------
```

## Steps to reproduce

I don't know that there are any per se.  It just started happening out of the blue.

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #6522

TODO measure performance impact

changelog: TODO

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt [hunk 1]
```diff
@@ -426,18 +426,10 @@ private fun <T : RsNamedElement> RsItemsOwner.getExpandedItemsWithName(
...
-                if (itemInner.name == name) {
-                    result.add(itemInner)
-                }
-            }
-        } else {
-            if (item.name == name && psiClass.isInstance(item)) {
-                @Suppress("UNCHECKED_CAST")
-                result.add(item as T)
-            }
+    for (item in expandedItemsCached.named.entriesWithName(name).values.flatten()) {
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt:429]: Looks like now we can use this simple implementation:
```kt
private inline fun <reified T : RsNamedElement> RsItemsOwner.getExpandedItemsWithName(name: String): List<T> =
    expandedItemsCached.named[name]?.filterIsInstance<T>() ?: emptyList()
```

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsItemsOwner.kt [hunk 2]
```diff
@@ -112,7 +126,7 @@ class RsCachedItems(
     val namedImports: List<CachedNamedImport>,
     val starImports: List<CachedStarImport>,
     val macros: List<RsMacro>,
-    val rest: List<RsItemElement>
+    val named: MutableMap<String, SmartList<RsItemElement>>
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsItemsOwner.kt:129]:
```suggestion
    val named: Map<String, List<RsItemElement>>
```

File: src/main/kotlin/org/rust/lang/core/resolve/ItemResolution.kt [hunk 3]
```diff
@@ -105,13 +107,13 @@ fun processItemDeclarations(
...
-                    }
-                }
-            }
+//            is RsForeignModItem -> if (Namespace.Values in ns) {
+//                for (child in item.stubChildrenOfType<RsNamedElement>()) {
+//                    if (child is RsFunction || child is RsConstant) {
+//                        if (processor(child)) return true
+//                    }
+//                }
+//            }
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/resolve/ItemResolution.kt:116]:
```suggestion

```

**Review by dima74 — APPROVED** [review]

**Comment by vlad20012** [comment]
bors r=dima74

**Pull request merged**